### ISSUE TITLE
Use `atts_since` in the replication algorithm

### ISSUE THREAD

**Issue by nolanlawson** [pouchdb/pouchdb#2674]
Currently we don't do this, which I guess means we read in every attachment for every revision? Bad for perf.

From the [TouchDB wiki](https://github.com/couchbaselabs/TouchDB-iOS/wiki/Replication-Algorithm):

> Also use the “?atts_since” query parameter to pass a list of revisions that the target already has, so the source can optimize by not including the bodies of attachments already known to the target.

**Comment by natcohen** [comment]
Any plan to fix this soon?

**Comment by nolanlawson** [comment]
None for me, hence the "help wanted" label. :)

**Comment by natcohen** [comment]
OK... Who can help? I can have a look at the algorithm so in which file is it?

**Comment by nolanlawson** [comment]
`lib/replicate.js` for the replication algorithm, and `lib/adapters/http/http.js` to pass the parameter to the http adapter.

**Comment by natcohen** [comment]
Unfortunately, I won't have time to help... hope someone else can take care of that. I think it's an important bug because it exponentially increases the size of couchdb which in the end represents money (cloudant/iriscouch)

**Comment by yaronyg** [comment]
I'm working on this now. Right now I'm just trying to figure out how our existing attachment code works so I can figure out what to stick in where.

**Comment by willholley** [comment]
@yaronyg I also went down that path (see https://github.com/pouchdb/pouchdb/tree/2674-atts-since) but came to the same conclusion as you - revpos is only required if you want pouchdb to support atts_since as a parameter to local document GET calls (i.e. when the replication source is local).

I created another branch which just adds support for passing atts_since through the http adapter (https://github.com/pouchdb/pouchdb/tree/2764-atts-since-http-only). This also contains functions can collect the set of local revisions for a given document - that could then be passed to atts_since during single document GETs (as is done by https://github.com/cloudant/sync-android/blob/43a4735c7f2356ff049ffc2c67e5babf9b9900e0/sync-core/src/main/java/com/cloudant/sync/replication/BasicPullStrategy.java#L395). However, PouchDB has to deal with browser URL length limits (see https://github.com/pouchdb/pouchdb/blob/master/lib/replicate.js#L227) - I'm not quite sure how to handle hitting the limit in this case given that we can't batch this as we do with missing revs.

**Comment by yaronyg** [comment]
I wanted to drop a quick note that I think there is potentially an easier way to implement atts_since. I need to finish writing some tests against CouchDB (that we will be able to re-use with PouchDB) to confirm my suspicions about how atts_since actually works. But first I need to submit a PR for supporting AJAX on GET requests in order to enable my tests. :)

**Comment by natcohen** [comment]
@yaronyg What kind of test exactly? Do you have any idea how much time it'll take?

**Comment by yaronyg** [comment]
@natcohen sorry for the delay. Unfortunately while fixing atts_since is part of my job, it's not currently my highest priority work item. We are bringing on some new people to the team and there are some core architecture issues I have to solve. So I don't expect to really get back to this until next week at the earliest.

Where I am right now is just trying to understand what CouchDB does with atts_since in various edge case scenarios. For example, I tested what happens if you create a doc with an attachment, create an update without the attachment and then add the attachment back in. If you use atts_since from the version before the delete will it remember the old value and return a stub?

I am also testing what happens when you have conflicts and attachments are at various places in the conflict tree.

The reason for writing these as tests is because I will be able to re-use all of this code when I try to implement atts_since to test our own behavior.

My goal is to get to the point where we have "proper" atts_since support on GET requests.

My next step would then be to put in support for using atts_since during our pull replication.

**Comment by natcohen** [comment]
@yaronyg don't you work for Microsoft? Is Microsoft officially involved in this project?

Sorry if it's too personal but I was just curious...it would be great to have a company supporting the project (like angularjs with Google, node-webkit with Intel, etc...)

**Comment by yaronyg** [comment]
@natcohen - I do work for Microsoft and the work I'm doing for PouchDB is part of my day job. But Microsoft the gestalt entity doesn't know it's supporting PouchDB. Rather it's supporting the [Thali Project](http://www.thaliproject.org) which depends on PouchDB. Currently we have three devs working on JS. @sreesharp just joined the Thali team and is coming up to speed. @mpodwysocki who is working on an ACL framework for PouchDB (he'll be posting to the mailing list about it soon). And me. At a minimum we have to seriously improve attachment perf (hence this bug) because that's a show stopper for us.
   But what would "support" look like to you from Microsoft? In other words, what can we do to help?

**Comment by natcohen** [comment]
@yaronyg Well, I'm a student at MIT and use this library for a project (unfortunately I don't have time to contribute...). My only concern is will it still be supported in 5 or even 10 years? Like any open source project (jquery, linux, etc...) it needs a strong team/community behind it to keep it up to date.

Fortunately, PouchDB has currently strong contributors but what if 3-4 of them leave the boat? What will happen? As I said, Google supports AngularJS so we know by using AngularJS most chances are that it will be supported for years (that's also why this framework gained popularity). Same for CouchDB  which is supported by Apache and indirectly by IBM through Cloudant...

How Microsoft could help? I guess as a non-contributor, I'm not the best person to answer but as a user all what I need is a guarantee the libraries I'm using won't die in the near future. I think how AngularJS is managed (community & private company) is probably the best for an open-source but I know Microsoft's philosophy is not very oriented toward open-source projects...

**Comment by nolanlawson** [comment]
@natcohen PouchDB has been supported pretty consistently for the past 4.5 years... That being said, I think what we are really trying to do (or at least what I'm trying to do) is build a platform, not a library.

More and more, I've found that I can make PouchDB meet whatever demands I may have by writing plugins and adapters. For instance, there's already work being done on [a plugin for AsyncStorage](https://github.com/nolanlawson/pouchdb-async-storage) to support React Native. I've also been working on [a websocket plugin](https://github.com/nolanlawson/socket-pouch).

These days, the core PouchDB features that remain to be implemented all revolve around performance (attachment performance, map/reduce performance, replication performance). The core code Just Works&trade;. That's a great place to be for any open-source project. :)

**Comment by yaronyg** [comment]
@natcohen  You are asking, in my opinion, the right questions. As a user you want to know - should you be investing in PouchDB?

In Microsoft's case, you're also right, we aren't terribly oriented to open source. [Here](http://www.microsoft.com/en-us/openness/default1.aspx#projects) is a list (incomplete I might add) of the tiny work Microsoft does with Open Source, only about 390 open source projects are listed there. So we obviously need to improve. :) 

But seriously, Microsoft is fully committed to open source and is a major contributor both to its own projects (like the recently open source .net framework) as well as to other projects like Redis, CouchDB, Mongo, Node, etc. So the issue really isn't so much Microsoft's commitment to Open Source, as its Microsoft's relationship to PouchDB.

What makes PouchDB different than AngularJS is that AngularJS *is* a Google project. In other words, it's owned by Google. So unless you want Microsoft to somehow "buy" PouchDB I don't think you want Microsoft to have an AngularJS style relationship with PouchDB. 

Even IBM and CouchDB is a hard comparison because CouchDB is part of the Apache Foundation. There are many positives to the Apache Foundation, amongst them that it makes the kind of guarantees and has the kind of financial basis that lets it do an awesome job making the open source trains run on time. But there is a real cost for this guarantees, to see that cost take a look at the developer overview of the [Apache Foundation Process](http://www.apache.org/dev/)

So if PouchDB isn't going to "sell" itself to some company, join a foundation like Apache or form its own foundation then it's hard for companies to make big investments. They can't be sure where things will go!

So at that point what you have to look at is - how healthy is PouchDb's community? Not just how often do they release. But how deep is their pool of talent? They have 3 core contributors. They have 8 contributors with over 50 commits. They have 140 contributors over all. That is... amazing! Just in the last year they have had 42 committers. If you look [here](http://redmonk.com/dberkholz/2013/04/22/the-size-of-open-source-communities-and-its-impact-upon-activity-licensing-and-hosting/) you will see that only 1.2% of projects have that many committers in a year. So that speaks to both a deep and wide pool of people actively involved with PouchDB.

Now is that a guarantee that PouchDB will be alive and well 3 or 4 years from now? Nobody can make you such a guarantee. Even if you paid for support for that long, companies (much less open source projects) go bankrupt. But given how long it has been around and how active it is it seems more likely that PouchDB will be doing well 3-4 years from now than most projects.

**Comment by nolanlawson** [comment]
@yaronyg Well said. Thank you :)

**Comment by natcohen** [comment]
@yaronyg Did you have a chance to write your tests?

**Comment by yaronyg** [comment]
@natcohen Sorry I was off on Paternity leave for a month and now I'm working on getting out the first story for Thali. Unfortunately it is going to be a while before we get back to the synchronization perf issues. We first have to get our basic infrastructure working. And when we do deal with synchronization perf improvement we aren't fully sure yet how we will handle it. There are a number of possible avenues.

1. Add in atts_since and mime/multipart. support & teach the replicator to use it,
2. Add mime/multipart to _bulk_get & teach the replicator to use it or
3. Take @nolanlawson 's work on streaming and add binary support to it

I'm honestly not sure at this point which one we will choose. Only that we will discuss the trade offs with the group. But as I said, right now we just have to get to the point where we can get our basics working. You can see our stories [here](http://thaliproject.org/stories)

**Comment by nolanlawson** [comment]
Welcome back! :)

I already worked on `multipart/mixed` (downloading). It's not going to be possible. E.g. see https://github.com/pouchdb/pouchdb/issues/2832#issuecomment-104982175. `multipart/related` (uploading) is done though, and published in 3.6.0.

`bulkGet()` is in progress and already largely implemented. I'm just waiting on CouchDB to merge, so I can start testing it proper.

If you could, I would focus on atts_since and revpos. There has been zero work done on it, and it would be a big win. :)

**Comment by yaronyg** [comment]
@nolanlawson In regards to multipart/mixed is there a plan to teach the replicator to ask for attachments separately?

Once we get to replication performance I'm happy to take my lead from you. But we aren't going to get to attachment perf until story 0.0.3.0 and we haven't gotten passed story -1 yet. :( See [here](http://thaliproject.org/stories) for what I'm talking about. So it's going to be a bit before we get there.

In fact right now we are spending a big chunk of our time just figuring out how to get Express-PouchDB to run happily on JXCore on top of Android and iOS. We are running into tons of small issues getting Express-PouchDB working on a mobile platform. None of this is Express-PouchDB's fault. It's just a bunch of small differences between Node.js and JXCore/mobile platforms that we have to work through.

**Comment by daleharvey** [comment]
We fetch attachments separately currently, been seeing some issues with the number of requests being uncapped but will likely be introducing a configurable functionality to specify the number of http requests

**Comment by natcohen** [comment]
@yaronyg Congrats and welcome back!

**Comment by yaronyg** [comment]
@natcohen Thanks!!!!

**Comment by lukevanhorn** [comment]
Hello, I've recently joined @yaronyg's team and I am continuing his work on this enhancement!

**Comment by nickcolley** [comment]
That's great to hear! @lukevanhorn let us know if you need any help :+1:

**Comment by nolanlawson** [comment]
@lukevanhorn Cool, if you need help, we're usually pretty responsive in the #pouchdb IRC channel. :)

**Comment by natcohen** [comment]
@lukevanhorn Welcome! Do you have any schedule to fix this issue?

**Comment by natcohen** [comment]
Still no plan to improve this?

**Comment by nolanlawson** [comment]
@natcohen This is an unfunded open-source project; stuff gets done when people implement it. :)

**Comment by janl** [comment]
Heya, just a heads up, @gr2m and I are looking into this now.

**Comment by natcohen** [comment]
Great! Good luck...

**Comment by janl** [comment]
# Implementing `atts_since` is not the right approach to fix this issue

## Story time!

CouchDB APIs used for replication generally have two modes: JSON and [multipart/mixed](https://en.wikipedia.org/wiki/MIME).

In the early days, CouchDB only did JSON APIs. It’s easy to implement in CouchDB and elsewhere, so that came first and a lot of clients implement it.

Later, CouchDB added multipart/mixed support, because in JSON-mode, CouchDB attachments were encoded in [base64](https://en.wikipedia.org/wiki/Base64) and that leads to a whole lot of inefficiencies (larger data size, mostly).

However, multipart/mixed decoding is harder to implement, so not all clients use this, including PouchDB. But since PouchDB doesn’t want to pay the base64-tax, it splits up fetching a document into 1+N requests, N for the number of attachments in the doc.

## The Problem

Now, `atts_since` in JSON mode (the only one that PouchDB understands, to my knowledge) only avoids sending the base64-encoded attachments that the target already has as specified in `?atts_since=["list-of-revisions-that-exist-on-the-target"]`. `atts_since` in multipart/mixed mode also avoids sending already-existing attachments, but PouchDB (again, afaik), doesn’t know how to read this. That is: **`atts_since` is solving a problem that PouchDB doesn’t have.**

To illustrate PouchDB’s behaviour, given this doc in the source database (over `http`):

```
{
    "_id": "foo"
    "_rev": "2-bsdfbeefcake",
    "_attachments": {
         "file1.jpg": {
            ...
         },
         "file2.jpg": {
            ...
         }
    }
}
```

When encountering this doc (with the `http` adapter), PouchDB will issue three requests:

1. `GET /database/foo?rev=2-bsdbeefcake`
2. `GET /database/foo/file1.jpg?rev=2-bsdbeefcake`
3. `GET /database/foo/file2.jpg?rev=2-bsdbeefcake`

and then stores things in the target. The replicator in PouchDB uses the `getDocs()` function on the `http` adapter. The function itself though doesn’t know anything about replication, it just fetches a doc, if so told, and all its attachments in separate requests (it also powers the primary `db.get()` function)).

So if have this scenario:

1. Document `foo` has one attachment `file1.jpg` at revision `1-asdf00dc0ma` in CouchDB.
2. PouchDB starts replication from CouchDB to a local CouchDB
  - PouchDB requests:
    1. `GET /database/foo?rev=1-asdf00dc0ma`
    2. `GET /database/foo/file1.jpg?rev=1-asdf00dc0ma`
3. Document `foo` gets another attachment `file2.jpg` in CouchDB.
4. PouchDB replicates again (or is in `live` mode):
  - PouchDB requests: 
    1. `GET /database/foo?rev=2-bsdbeefcake`
    2. `GET /database/foo/file1.jpg?rev=2-bsdbeefcake`
    3. `GET /database/foo/file2.jpg?rev=2-bsdbeefcake`

As I hope is clear by now, request `b.` is superfluous here, because we already have that data in the target from the first replication.


## The Solution

In order to fix this, we need two things:

1. a way to give `getDocs()` access to the target database, if it is replicating.
2. a way to tell `getDocs()` that it should only fetch an attachment from the source database (if it is over `http`), if it doesn’t exist locally.
3. a way to look up if the correct version of an attachment exists in the replication target.

We (@gr2m) have prepared a patch that illustrates a potential implementation. It is very rough and optimised for minimal changes, not for proper code architecture. We’d like to hear from the PouchDB team if, and how to do this better.

Please ignore the whitespace changes in the tests, only the changes in `src` are relevant: https://github.com/gr2m/pouchdb/commit/7ef5849bf1f94398c7e397dccbefb144302004a4

You will find the following:

1. passing a reference to the target database into `getDocs()` during replication:
  - https://github.com/gr2m/pouchdb/commit/7ef5849bf1f94398c7e397dccbefb144302004a4#diff-3b1908d46435f961119df0ed4762ec7fR35
  - https://github.com/gr2m/pouchdb/commit/7ef5849bf1f94398c7e397dccbefb144302004a4#diff-1b1f318b599e364220cdb8ac006c238bR151

2. a way to check for the existence of an attachment on the target within `getDocs()`, and if it doesn’t exist, continue on the existing code path of fetching the attachment from the source:
  - https://github.com/gr2m/pouchdb/commit/7ef5849bf1f94398c7e397dccbefb144302004a4#diff-99a0c1529f05283f3fd14198f302c176R446

3. the look-up of the attachment on the source and identification of right version:
  -  https://github.com/gr2m/pouchdb/commit/7ef5849bf1f94398c7e397dccbefb144302004a4#diff-3b1908d46435f961119df0ed4762ec7fR43

When applied, you will see that PouchDB will not fetch attachments from the source if they already exist in the target. In the example above, instead of the request `b.` in the second replication, PouchDB checks if that attachment already exists in the target, and if yes, proceeds to only fetch the other attachment.

We’re looking forward to your feedback! <3

And let us know if you have any questions!

**Comment by nolanlawson** [comment]
This sounds like an elegant solution to `atts_since` problem. Some quick comments:

1. Yes, PouchDB does not understand multipart/mixed. This is due to limitations in the browser itself and in CouchDB. Namely, CouchDB always serves multipart attachments gzip-encoded, and it's prohibitively expensive for us to include a JS-based GZip decoder as part of PouchDB (there is no browser API for this).
2. Note that we also still plan on adding an `inline` option (https://github.com/pouchdb/pouchdb/issues/3913), which would make an HTTP adapter prefer base64 inlined attachments rather than fetching binary attachments. However this seems like it could be compatible (or at least orthogonal) with your implementation.
3. @gr2m's implementation is great in principle, although I do agree the code could be cleaned up a bit. In particular I think maybe we could get rid of attaching a function to the `opts` object and instead attach it to the http adapter itself. Then the replicator could do all the "if http, do a local fetch, then fall back to a normal fetch." Also I'm not sure the extra http function is even required; it seems the new `fetchAttachment()` is almost exactly the same as the existing `api.getAttachment()` in `http.js`, although correct me if I'm wrong.

Anyway thanks a ton for both of your work on this; this has been a stagnant issue for a long time, and I know it's a much-requested feature to improve replication performance. :)

**Comment by janl** [comment]
@nolanlawson thank you for your in-depth comments! <3

As for 1. (and sorry for the derail) I remember folks hacking the `img` tag’s PNG decoder to do “native” gzip in browsers. It is definitely a hack, but it might be an option there?

I think 2. is definitely also good to have, in addition to this :)

As for 3. I’ll work with @gr2m on addressing your suggestions.

**Comment by nolanlawson** [comment]
@janl For the `<img>` hack, I'd kind of rather not, especially since that would then require a Node/browser switcheroo and would maybe still have bad performance (UI blocking?) as well as not working in a web/service worker (the real dealbreaker).

Once # 3 is cleaned up, I think the suggestion is totally reasonable and probably easy to integrate. In an ideal world, we may even be able to isolate the change to only the replicator (and thus avoid adding new APIs to any adapters).

**Comment by janl** [comment]
@nolanlawson that sounds sensible, thanks for the perspective! — aside: if it isn’t under way, we should look at having native deflate in browsers ;)

**Comment by gr2m** [comment]
see PR here: https://github.com/pouchdb/pouchdb/pull/5018/

**Comment by jo** [comment]
a little off topic, but related: I implemented `revpos`: https://github.com/pouchdb/pouchdb/pull/5037

**Comment by jo** [comment]
As it turns out that the [digest approach](https://github.com/pouchdb/pouchdb/pull/5018) might not work because of [unreproducible digests](https://issues.apache.org/jira/browse/COUCHDB-2252) we need to find an other way to address the issue.

_**tl;dr** I think `atts_since` is still an option. We can implement it client side._

@janl argues that, since PouchDB uses neither `multipart/mixed` requests nor JSON request with inlined attachments but instead makes multiple requests for the doc and each of its attachments,  `atts_since` does not solve our problem.
The solution is simple: we implement `atts_since` in PouchDB, client side: first we request the doc and then we compare the `revpos` properties with the given revisions in the `atts_since` parameter to decide which attachments needs to be fetched.

Lets take @janl s example to illustrate how this would work:

```
{
    "_id": "foo"
    "_rev": "2-bsdfbeefcake",
    "_attachments": {
         "file1.jpg": {
            "revpos": 1,
            ...
         },
         "file2.jpg": {
            "revpos": 2,
            ...
         }
    }
}
```

The replicator in PouchDB uses the `getDocs()` function on the `http` adapter and passes the revision of the current document as `atts_since: ['1-asdf00dc0ma']`. The function itself now first fetches the doc, inspects its attachment stubs and fetches only attachments which have `revpos > 1` resulting in the two requests:

1. `GET /database/foo?rev=2-bsdbeefcake`
2. `GET /database/foo/file2.jpg?rev=2-bsdbeefcake`

So we now have this overall scenario:

1. Document `foo` has one attachment `file1.jpg` at revision `1-asdf00dc0ma` in CouchDB.
2. PouchDB starts replication from CouchDB to a local CouchDB
  - PouchDB requests:
    1. `GET /database/foo?rev=1-asdf00dc0ma`
    2. `GET /database/foo/file1.jpg?rev=1-asdf00dc0ma`
3. Document `foo` gets another attachment `file2.jpg` in CouchDB.
4. PouchDB replicates again (or is in `live` mode):
  - Replicator requests: 
    1. `GET /database/foo?rev=2-bsdbeefcake`
    2. `GET /database/foo/file2.jpg?rev=2-bsdbeefcake`

And we're fine.

What do you think?

**Comment by yaronyg** [comment]
Imagine that Machine A and Machine B both have a database foo and both, independently, create a doc with the ID "bar". The contents of the doc are different however. As a result each has a rev of the form say 1-A and 1-B. Just for fun, let's say that each of the docs has an attachment, so both attachments have revpos 1.

When the two databases try to sync there will be a conflict. But the point is that if say DB A just wants to replicate DB B's content it has to look at revpos in the context of the associate rev. In that _rev 1-A + revpos 1 is different than _rev 1-B + revpos 1.

And can PouchDB handle a situation where say DB A isn't sure how to resolve the conflict and so it copies over 1-B. Can it actually copy it over as 1-B or does it have to change the _rev to 2-B just to make sure that the local revs stay in a linear order? So now 2-B on DB A is the same as 1-B on DB B?

In other words is the sequence number part of rev only local in the same way that the over all DB sequence number is only local?

Sorry for the newbie questions but I was never really clear on this.

**Comment by lukevanhorn** [comment]
fyi, this may or may not be helpful, but I merged my fork with the latest upstream. May be worth a glance to see if I was going in the right direction.

https://github.com/lukevanhorn/pouchdb

changes are in:
adapter.js
adapters/websql/index.js
deps/bulkGetShim.js
deps/docs/updateDoc.js

I recall being slightly derailed in testing, but I don't recall the exact reasons things weren't working as expected with pouchdb->pouchdb sync. Sorry, I can't spend more time on it now.
